c10/core/RefcountedDeleter.cpp

```diff
@@ -0,0 +1,73 @@
+#include <c10/core/RefcountedDeleter.h>
+
+#include <mutex>
+
+namespace c10 {
+
+void refcounted_deleter(void* ctx_) {
+  RefcountedDeleterContext& ctx =
+      *reinterpret_cast<RefcountedDeleterContext*>(ctx_);
+  ctx.refcount--;
+  if (ctx.refcount == 0) {
+    ctx.other_ctx = nullptr;
+    delete &ctx;
+  }
+}
+
+std::mutex replace_data_ptr_mutex;
```

Comment:
Single global mutex shared with everyone? 🤔

Reply:
Hmm, but there is not any other logical place to put it.

Reply:
Maybe there is a better way to do this. If two threads are calling `maybeApplyRefcountedDeleter` on two different storages, neither of them actually needs to wait. Threads only need to wait if another thread is operating on the same storage. We could have some sort of global atomic table that says whether a different thread is currently operating on a given storage or not. That would improve performance, but I'm not sure at the moment how to implement something like that

And I'm not sure if performance is an issue here anyway. I assumed that if someone is sharing storages between different threads in mulitpy, they would just do all the sharing at the beginning of the process as an initialization step. But I could be wrong--I don't know how people use multipy

Reply:
Yeah I think I have made my peace with this.